### PR TITLE
Added tests about permissions and public upload

### PR DESCRIPTION
**Pull request by SergioBertolinSG**
Test for https://github.com/owncloud/QA/issues/301
And a current failing case. cc @PVince81

### CODE REVIEW

File: build/integration/features/sharing-v1.feature [hunk 1]
```diff
@@ -976,3 +976,37 @@ Feature: sharing
...
+    And As an "user0"
+    And user "user0" created a folder "/test"
+    And folder "/test" of user "user0" is shared with user "user1"
+    And As an "user1"
+    And creating a share with
+      | path | /test |
+      | shareType | 3 |
+      | publicUpload | false |
+    When Updating last share with
+      | publicUpload | true |
```
Comment (PVince81) [inline build/integration/features/sharing-v1.feature]: This looks legitimate. At line 1002 the share is read-write so it's acceptable that the recipient can also make the link share read-write.

Did you mean to make the local share read-only ?

Comment (SergioBertolinSG) [inline build/integration/features/sharing-v1.feature]: Yes, you're right. I'll change it.

File: build/integration/features/sharing-v1.feature [hunk 2]
```diff
@@ -976,3 +976,37 @@ Feature: sharing
...
     And etag of element "/PARENT" of user "user0" has not changed
 
+  Scenario: Increasing permissions is allowed for owner
+    Given As an "admin"
+    And user "user0" exists
+    And user "user1" exists
+    And group "new-group" exists
+    And user "user0" belongs to group "new-group"
+    And user "user1" belongs to group "new-group"
+    And Assure user "user0" is subadmin of group "new-group"
```
Comment (SergioBertolinSG) [inline build/integration/features/sharing-v1.feature]: Perhaps not, it comes from here https://github.com/owncloud/core/issues/25542#issuecomment-243061790

**Comment by PVince81** [comment]
Pfff, this API thing is a real mess. Some checks are done in the share manager, others are done only in the OCS controller...

**Comment by PVince81** [comment]
Basically the `createShare` method in the OCS controller doesn't check if permissions can be increased, but a check is done in the sharemanager.
However for `updateShare` it is done in the OCS controller but not in the share manager.

@SergioBertolinSG can we also have integration tests for creating link shares as reshares for a read-write and read-only received share ?

**Comment by PVince81** [comment]
Debugging in the `updateShare` case, apparently the share manager calls `$share->getNode()->getPermissions()` which returns 31. These are the permissiosn from the POV of the owner, not the recipient.

But weird thing that this check is correct when creating shares...

**Comment by PVince81** [comment]
Ok, got it. The problem is as follows: the `Node` that was set to the `Share` object to be modified is different depending if create or update:
- for `createShare`: the Node is set to "/user1/files/test", the received folder
- for `updateShare`: since the owner is "user0", the node sets itself to "/user0/files/test". And because of this, the permissions returned for the share's node is different, the ones from the owner's POV not the recipient. However when creating shares, we always need the permissions from the POV of the recipient.

So need to look into whether we need to "fix" the node in the share manager, or adjust the node in the OCS controller code. It looks like the logic would let us discard the permissions validation logic from the OCS controller and only rely on the one in the manager, which would be ideal.

**Comment by PVince81** [comment]
Looks like for the share manager we need to somehow try and get the node with the POV of the user who is specified in "getSharedBy()" instead of relying on the node that was specified in the share itself as it might be inaccurate.

**Comment by PVince81** [comment]
Some POC here: https://github.com/owncloud/core/pull/26639/commits/fe38088a3bbbc0c4d203fa1eb4f9eee3cc43a60e

Would be good to add many more integration tests to make sure everything is covered (including reshare-creation perm denied)
Will also add unit tests for the share manager change.

**Comment by SergioBertolinSG** [comment]
> @SergioBertolinSG can we also have integration tests for creating link shares as reshares for a read-write and read-only received share ?

You mean removing permission to share?

**Comment by PVince81** [comment]
> You mean without permission to share?

I didn't mean that. But you're right that it needs to be added as well.

**Comment by PVince81** [comment]
For now I'll focus on the quick fix. The clean but more risky fix will be done for 9.2 separately: https://github.com/owncloud/core/pull/26684 (more risky because it will affect more cases that we need to make sure we cover as well)

File: build/integration/features/sharing-v1.feature [hunk 3]
```diff
@@ -976,3 +976,54 @@ Feature: sharing
...
+    And As an "user0"
+    And user "user0" created a folder "/test"
+    And folder "/test" of user "user0" is shared with user "user1" with permissions 31
+    And As an "user1"
+    And creating a share with
+      | path | /test |
+      | shareType | 3 |
+      | publicUpload | false |
+    When Updating last share with
+      | publicUpload | true |
```
Comment (PVince81) [inline build/integration/features/sharing-v1.feature]: Can we also have the same tests with the "permissions" attribute ? @SergioBertolinSG

Comment (SergioBertolinSG) [inline build/integration/features/sharing-v1.feature]: Sure.

**Comment by SergioBertolinSG** [comment]
> @SergioBertolinSG can we also have integration tests for creating link shares as reshares for a read-write and read-only received share ?

With sharing permission, these cases are covered in 
'Scenario: Adding public upload to a shared folder as recipient is allowed with permissions' and 'Scenario: Adding public upload to a read only shared folder as recipient is not allowed'
aren't they?

Without sharing permissions there is 'Scenario: User is not allowed to reshare file' and 'Scenario: User is not allowed to reshare file with more permissions'. They reshare like regular shares, Do we need different tests using public links shares?

**Comment by PVince81** [comment]
> Do we need different tests using public links shares?

Yes, that would be good as it's a slightly different code path.

**Comment by PVince81** [comment]
Seems Jenkins env is broken, I resubmit the PR with your commits squashed here: https://github.com/owncloud/core/pull/26691

**Pull request closed without merging**